api/v1alpha1/ratelimitfilter_types.go

```diff
@@ -0,0 +1,133 @@
+// Copyright Envoy Gateway Authors
+// SPDX-License-Identifier: Apache-2.0
+// The full text of the Apache license is available in the LICENSE file at
+// the root of the repo.
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// +genclient
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:printcolumn:name="Age",type=date,JSONPath=`.metadata.creationTimestamp`
+
+// RateLimitFilter allows the user to limit the number of incoming requests
+// to a predefined value based on attributes within the traffic flow.
+type RateLimitFilter struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// Spec defines the desired state of HTTPRoute.
+	Spec RateLimitFilterSpec `json:"spec"`
+}
+
+// +kubebuilder:object:root=true
+
+// RateLimitFilterList contains a list of RateLimitFilter resources.
+type RateLimitList struct {
+	metav1.TypeMeta `json:",inline"`
+	metav1.ListMeta `json:"metadata,omitempty"`
+	Items           []RateLimitFilter `json:"items"`
+}
+
+// RateLimitFilterSpec defines the desired state of RateLimitFilter
+// +union
+type RateLimitFilterSpec struct {
+	// Type decides the scope for the RateLimits.
+	// +unionDiscriminator
+	Type RateLimitType `json:"type"`
+	// Rules are a list of RateLimit matchers and limits.
+	//
+	// +kubebuilder:validation:MaxItems=16
+	Rules []RateLimitRule `json:"rules"`
+}
+
+// RateLimitType specifies the types of RateLimiting.
+// Valid RateLimitType values are:
+//
+// * "Global"
+//
+// +kubebuilder:validation:Enum=Global
+type RateLimitType string
+
+const (
+	// In this mode, the rate limits are applied across all Envoy proxy instances.
+	RateLimitTypeGlobal RateLimitType = "Global"
+)
+
+// RateLimitRule defines the semantics for matching attributes
+// from the incoming requests, and setting limits for them.
+type RateLimitRule struct {
+	// +optional
+	// +kubebuilder:validation:MaxItems=8
+	Matches []RateLimitMatch `json:"matches,omitempty"`
```

Comment:
"matches" as a name also might be what is tripping people up here

I assume this is meant to express that rate limits should be counted separately for requests with differing attributes, e.g. for a HTTPRoute with a rule for `/foo` that has a RL filter, you may want to express that a limit of 10 rps is set separately on requests with `x-user-id: one` and `x-user-id: two`, or following the examples in this doc you want to specify that a rate limit is only set for `x-user-id: one`

I tend to not think of this so much as a "matching" condition but rather maybe a "discriminator" property of the rate limit

Another feature of this doc that might be tripping people up is that the "exact" header match defined here and in the examples is pretty simple and easily expressed in an HTTPRoute rule match, so it feels like it doesn't need to be expressed in this way. To make things more convincing it might be useful to take the examples further to their likely conclusion in terms of different "match types" (to use the existing terminology).

Reply:
the naming for **some** of the _what to match and how_ are identical to HTTPRoute (like `matches` and `Exact` to) so reuse existing word definition reducing cognitive load on the user

Reply:
I would argue the "matches" field as it is named today (not sure its worth coming up with something different than "exact", that wasn't the intent of my comment) expresses an entirely different concept from route matching so should not use the same term, but that might just be personal preference

route match conditions are ORed or ANDed together to forward traffic to a set of backends, while matches/descriptors/discriminators in the rate limiting context do not necessarily combine to "counting against one rate limit" but can also be used to ensure the same rate limits are counted separately for subsets of traffic that match the route

its not quite "many conditions" -> one outcome (forward to backends), its more "many RL conditions" -> many RL buckets

Reply:
the ANDing logic within a single match rule (multiple conditions) and ORing across `matches` is same in RL and HTTPRoute

Reply:
I'm not saying the logical operations are totally different, but the outcomes are: HTTPRoute matches -> forwarding to a backend, RL "matches" -> can end up with one or more RL buckets so it is a bit confusing to me to me personally to use the word "matches", I prefer something to the effect of "discriminators" but whatever works

Reply:
from a user perspective, its the same -  RL "matches" -> if limit reached -> limits
what is happening under the hood is a little more complex, but the this user facing API doesnt need to reflect it.
wrt naming, in the past, I had used the word `dimension` to define this term, if we named it `dimension` here, the user would have to go through the docs in detail to understand its usage, by using `matches`, the user should be able to use this API faster since they are already aware of what it means,  since its same as the one in HTTPRoute